src/rust/engine/fs/store/src/snapshot_ops.rs

```diff
@@ -0,0 +1,731 @@
+          // `VFS` instances!
+          match &path_glob {
+            RestrictedPathGlob::Wildcard { wildcard } => {
+              assert_ne!(*wildcard, *DOUBLE_STAR_GLOB);
+              // This directory matched completely, without having to subset its contents
+              // whatsoever. We can avoid traversing its contents and return the node.
+              // NB: This line should be idempotent if the same directory node is added twice.
+              globbed_directories.insert(directory_path, directory_node.clone());
+            }
+            RestrictedPathGlob::DirWildcard {
+              wildcard,
+              remainder,
+            } => {
+              let mut subdir_globs: Vec<RestrictedPathGlob> = vec![];
+              if *wildcard == *DOUBLE_STAR_GLOB {
+                // NB: here we short-circuit several cases which would swallow up a directory
+                // without subsetting it.
+                match &remainder[..] {
+                  // ending in /**
+                  [] => {
+                    globbed_directories.insert(directory_path, directory_node.clone());
+                    continue;
+                  }
+                  // ending in /**/* or /**/**
+                  [single_glob]
+                    if *single_glob == *SINGLE_STAR_GLOB || *single_glob == *DOUBLE_STAR_GLOB =>
+                  {
+                    globbed_directories.insert(directory_path, directory_node.clone());
+                    continue;
+                  }
+                  _ => (),
+                }
+                let with_double_star = RestrictedPathGlob::DirWildcard {
+                  wildcard: wildcard.clone(),
+                  remainder: remainder.clone(),
+                };
+                subdir_globs.push(with_double_star);
+              } else {
+                assert_ne!(0, remainder.len());
+              }
+              match remainder.len() {
+                0 => (),
+                1 => {
+                  let next_glob = RestrictedPathGlob::Wildcard {
+                    wildcard: remainder.get(0).unwrap().clone(),
+                  };
+                  subdir_globs.push(next_glob);
+                }
+                _ => {
+                  let next_glob = RestrictedPathGlob::DirWildcard {
+                    wildcard: remainder.get(0).unwrap().clone(),
+                    remainder: remainder[1..].to_vec(),
+                  };
+                  subdir_globs.push(next_glob);
+                }
+              }
+              // Append to the existing globs, and collate at the end of this iteration.
+              let entry = todo_directories
+                .entry(directory_path)
+                .or_insert_with(Vec::new);
+              entry.extend(subdir_globs);
+            }
+          }
+        }
+      }
+
+      // 1c. Push any "dormant" contexts to traverse subdirs onto the dormant stack.
+      let dependencies: Vec<PathBuf> = {
```

Comment:
Still not sure what dormant or dependencies would mean here.

Reply:
I've edited the comment to remove this language.